### PR TITLE
WIP: Add ability for cache to do Cleanup

### PR DESCRIPTION
**Pull request by jmccann**
This is my stab at providing base functionality for cache plugins to be able to cleanup old archives.

### CODE REVIEW

File: storage/storage.go [hunk 1]
```diff
@@ -1,9 +1,19 @@
 package storage
 
-import "io"
+import (
+	"io"
+	"os"
+)
+
+type File struct {
```
Comment (donny-dont) [inline storage/storage.go]: It looks like FileInfo contains the actual Name. See https://golang.org/src/os/types.go?s=391:752#L6 so I don't think you need to make a new struct here. So for like s3 in the list you would create the FileInfo and add the name and add the modified time and then things would just work.

**Review by donny-dont — CHANGES_REQUESTED** [review]
Please verify we can do everything we want to with just FileInfo. If not we can go from there.

File: storage/storage.go (hunk 1 above)
Comment (jmccann) [reply storage/storage.go]: `FileInfo` does not contain the path, only the base name of the file with `Name()`.  That is why I created the `File` struct ... to keep the `Path` along with the `FileInfo` details.

Comment (donny-dont) [reply storage/storage.go]: Ah okay I misunderstood the struct. I don't like the name though. Maybe `FileEntry`?.

Also let me ask @bradrydzewski if there's something already around which is similar to what we want already in the library. Or maybe he would prefer something like a Walk function.

Comment (jmccann) [reply storage/storage.go]: I'll update from `File` to `FileEntry` and we can see what @bradrydzewski 's input is to this all in general.  :)

**Comment by bradrydzewski** [comment]
one idea might be to create a `cache.Flusher` interface, with a `cache.TimeFlusher` as the default implementation. With an optional `NoopFlusher` for individuals that don't want this functionality. These could get passed to cache in the constructor

```diff
type Cache struct {
	s storage.Storage
+	a archive.Archiver
+	f Flusher
}

-func New(s storage.Storage) (Cache, error) {
+func New(s storage.Storage, a archive.Archiver, f Flusher) Cache {
	return Cache{ s: s, a: a, f: f }
}
```

Note that I think we should also pass the archiver into the new function. That way the individual using the library has direct control over the compression algorithm being used, and can customize without asking for changes in the lib.

Since most people probably won't need to customize, we can probably provide the following function to set the following defaults:

```
func NewDefault(s storage.Storage) Cache {
	return New(s, new(archiver.TarArchiver), new(TimeFlusher))
}
```

**Comment by jmccann** [comment]
@bradrydzewski @donny-dont I tried to create a `Flusher`.  Not sure I did it entirely right.  I feel overall this is better but still not where we want it.  Unfortunately I'm still new to designing stuff so it's not coming naturally to me.  Give me enough direction and I can fix/clean this up more.  Thanks for the reviews!

Also, `archive/util` is no longer being used at this point.  Didn't remove it yet but if enough of this is looking good and we don't think we need that anymore let me know and I'll remove it.

**Comment by bradrydzewski** [comment]
> Not sure I did it entirely right

agreed, I'm starting to think flusher should not be used by the cache struct. Maybe it should be a separate thing, instead?

```Go
f := cache.NewFlusher(s)
f.Match = func(f *File) bool {
  return file.Timestamp().After(time.Now() - time.Days - 30) // pseudo code, won't work
}
f.Flush()
```

Just an idea

```GO
type Flusher struct {
  store Storage
  Match func(File) bool
}

func (f *Flusher) Flush() error {
  log.Infof("Cleaning files from %s", src)

  files, err := f.store.List(src)
  if err != nil {
    return err
  }

  for _, file := range files {
    if f.Match(file) {
      err := f.store.Delete(file)
      if err != nil {
        return err
      }
    }
  }

  return nil
}
```

**Comment by jmccann** [comment]
I removed `Flusher` from `Cache`.  It should be called directly by a plugin that wants to cleanup files.

I think I got close (or maybe accomplished) what you were describing @bradrydzewski ?

I'm feeling a lot better about how this looks personally.  Let me know of any other thoughts/comments.  If you guys think it's looking good then I'll try implementing for reals in the `drone-s3-cache` to see if it'll really really work.  😄

File: flusher/flusher.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+func TimeOp() dirtyFunc {
+	return func(file storage.FileEntry) bool {
+		// Do not delete directories
```
Comment (bradrydzewski) [inline flusher/flusher.go]: In this case you don't need to return a function, you can just do this:

```
func TimeOp(file storage.FileEntry) bool {
 ...
}
```

**Comment by bradrydzewski** [comment]
Looking great. I added one inline comment. The only other minor comment I have is that I would probably put flusher inside of the cache package. There is of course no hard rule when to put something in a separate package or not. I can't articulate why, but it feels like it belongs in cache :)

File: flusher/flusher.go (hunk 2 above)
Comment (jmccann) [reply flusher/flusher.go]: How do I then assign the function in `NewDefaultFlusher` @ https://github.com/drone/drone-cache-lib/blob/7966928e65d26567af5eded3326bd01316b031dc/cache/flusher.go#L23 ?

**Comment by jmccann** [comment]
I made the updates.

I added a question/comment to your inline comment.  I'll get that cleared up once I have an answer.  :)

Also, just curious how to handle testing between `cache_test.go` and `flusher_test.go`.  They use some of the same fixture functions.  Is it OK how it is and have them declared in `cache_test.go` and use in `flusher_test.go`?  Or should I put shared functions in a separate file?

**Comment by bradrydzewski** [comment]
> How do I then assign the function in NewDefaultFlusher

You can assign the function like this:

```
func NewDefaultFlusher(s storage.Storage) Flusher {
	return Flusher{ store: s, dirty: IsExpired }
}
```

Where the function is defined like this:

```
func IsExpired(file storage.FileEntry) bool {
    ....
    return true
}
```

Note that I would recommend the minor name change, just to be a bit more descriptive :)

> Also, just curious how to handle testing between cache_test.go and flusher_test.go. They use some of the same fixture functions. 

I think what you have now is pretty good.

I do think it could be improved by removing the fixtures all together and using a [filesystem interface](https://golang.org/pkg/net/http/#FileSystem) and a [mock](https://github.com/golang/mock) testing library. But I would say this could be explored at another time :)

**Comment by bradrydzewski** [comment]
awesome, thanks @jmccann !

**Pull request merged**